### PR TITLE
heal: calculate the number of workers based on  NRRequests

### PR DESCRIPTION
**Pull request by vadmeste**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license] (https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/global-heal.go [hunk 1]
```diff
@@ -173,6 +173,21 @@ func (er *erasureObjects) healErasureSet(ctx context.Context, buckets []string,
...
+	}
+
+	numHealers := uint64(1)
+	if info.NRRequests > 0 {
+		numHealers = info.NRRequests
+	}
+	if numHealers > 128 {
+		numHealers = 128
+	}
+
```
Comment (harshavardhana) [inline cmd/global-heal.go]:
```suggestion
	numHealers := uint64(4)
	if info.NRRequests > runtime.GOMAXPROCS(0) {
	      numHealers = runtime.GOMAXPROCS(0)/4
	} else {
	      numHealers = info.NRRequests/4
	}
	if numHealers == 0 || numHealers == 1 {
	       numHealers = 4
	}
	// allow overriding this value as well..
	if v := globalAPIConfig.getHealWorkers(); v > 0 {
	     numHealers = v
	}
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: internal/config/heal/heal.go [hunk 2]
```diff
@@ -103,6 +115,11 @@ var DefaultKVS = config.KVS{
 		Key:   IOCount,
 		Value: "100",
 	},
+	config.KV{
+		Key:        Workers,
+		Value:      "",
+		Deprecated: true,
```
Comment (harshavardhana) [inline internal/config/heal/heal.go]:
```suggestion
		Deprecated: true,
```

:P

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: internal/config/heal/heal.go (hunk 2 above)
Comment (harshavardhana) [reply internal/config/heal/heal.go]: @vadmeste you need to remove "Deprecated: true"

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by vadmeste** [comment]
> @vadmeste you need to remove "Deprecated: true"

The point is to make this variable hidden from users, until someone expliclity sets it

**Comment by harshavardhana** [comment]
> The point is to make this variable hidden from users, until someone explicitly sets it

For that we can't reuse this boolean, we should simply have this available to all users but document that it is not needed to be set without expert guidance via SUBNET

File: cmd/xl-storage.go [hunk 3]
```diff
@@ -244,6 +246,11 @@ func newXLStorage(ep Endpoint, cleanUp bool) (s *xlStorage, err error) {
 		diskIndex:  -1,
 	}
 
+	// Sanitaize before setting it
```
Comment (harshavardhana) [inline cmd/xl-storage.go]:
```suggestion
	// Sanitize before setting it
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/global-heal.go [hunk 4]
```diff
@@ -173,6 +153,28 @@ func (er *erasureObjects) healErasureSet(ctx context.Context, buckets []string,
...
+	if numHealers == 0 || numHealers == 1 {
+		numHealers = 4
+	}
+	// allow overriding this value as well..
+	if v := globalHealConfig.GetWorkers(); v > 0 {
+		numHealers = uint64(v)
+	}
+
+	jt, _ := workers.New(int(numHealers))
+
```
Comment (harshavardhana) [inline cmd/global-heal.go]: let's also update https://github.com/minio/minio/blob/e3fbcaeb729f5dd02f4b7c96c10b952c1d7dfe2a/docs/config/README.md?plain=1#L274

To show this configurable value

Comment (harshavardhana) [reply cmd/global-heal.go]: We have to say that this is Linux only optimization, for example on Windows and other OS we will use `number of cores / 4`

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/global-heal.go [hunk 5]
```diff
@@ -173,6 +153,28 @@ func (er *erasureObjects) healErasureSet(ctx context.Context, buckets []string,
...
+	}
+
+	var numHealers uint64
+
+	if numCores := uint64(runtime.GOMAXPROCS(0)); info.NRRequests > numCores {
+		numHealers = numCores / 4
+	} else {
+		numHealers = info.NRRequests / 4
+	}
+	if numHealers == 0 || numHealers == 1 {
```
Comment (harshavardhana) [inline cmd/global-heal.go]:
```suggestion
	if numHealers < 4 {
```

We expect this to be 4 minimum, unless user wants this to be some other value.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**